File: litho-rendercore-text/src/main/java/com/facebook/rendercore/text/TextStyle.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.rendercore.text;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.text.TextUtils;
import android.view.Gravity;
import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;
import androidx.core.text.TextDirectionHeuristicCompat;
import androidx.core.text.TextDirectionHeuristicsCompat;

public class TextStyle implements Cloneable {
  static final int UNSET = -1;
  @Nullable TextUtils.TruncateAt ellipsize;
  @Nullable CharSequence customEllipsisText;
  boolean includeFontPadding = true;
  int minLines = Integer.MIN_VALUE;
  int maxLines = Integer.MAX_VALUE;
  int minEms = UNSET;
  int maxEms = UNSET;
  int minTextWidth = 0;
  int maxTextWidth = Integer.MAX_VALUE;
  float shadowRadius = 0;
  float shadowDx = 0;
  float shadowDy = 0;
  int shadowColor = Color.GRAY;
  boolean isSingleLine = false;
  @ColorInt int textColor = Color.BLACK;
  @Nullable ColorStateList textColorStateList;
  int linkColor = Color.BLUE;
  int textSize = UNSET;
  boolean shouldAddSpacingExtraToFirstLine;
  float lineSpacingExtra = 0;
  float lineHeight = Float.MAX_VALUE;
  float lineHeightMultiplier = 1;
  float letterSpacing = 0;
  int textStyle = Typeface.DEFAULT.getStyle();
  @Nullable Typeface typeface;
  TextAlignment alignment = TextAlignment.TEXT_START;
  int breakStrategy = UNSET;
  int hyphenationFrequency = 0;
  int justificationMode = 0;
  TextDirectionHeuristicCompat textDirection;
  boolean clipToBounds = true;
  VerticalGravity verticalGravity = VerticalGravity.TOP;
  int highlightColor = Color.TRANSPARENT;
  int highlightStartOffset = UNSET;
  int highlightEndOffset = UNSET;
  int highlightCornerRadius = 0;
  boolean shouldLayoutEmptyText = false;

  boolean shouldTruncateTextUsingConstraints = false;
  int manualBaselineSpacing = Integer.MIN_VALUE;
  int manualCapSpacing = Integer.MIN_VALUE;
  float extraSpacingLeft = 0;
  float extraSpacingRight = 0;
  RoundedBackgroundProps roundedBackgroundProps = null;
  @Nullable String accessibilityLabel;

  public void setAccessibilityLabel(String accessibilityLabel) {
    this.accessibilityLabel = accessibilityLabel;
  }

  public void setShouldLayoutEmptyText(boolean shouldLayoutEmptyText) {
    this.shouldLayoutEmptyText = shouldLayoutEmptyText;
  }

  public void setShouldTruncateTextUsingConstraints(boolean shouldTruncateTextUsingConstraints) {
    this.shouldTruncateTextUsingConstraints = shouldTruncateTextUsingConstraints;
  }

  public void setManualBaselineCapSpacing(int manualBaselineSpacing, int manualCapSpacing) {
    this.manualBaselineSpacing = manualBaselineSpacing;
    this.manualCapSpacing = manualCapSpacing;
  }

  public TextStyle makeCopy() {
    try {
      return (TextStyle) super.clone();
    } catch (CloneNotSupportedException e) {
      throw new RuntimeException(e);
    }
  }

  public static TextStyle createDefaultConfiguredTextStyle(final Context context) {
    final TextStyle textStyle = TextStylesAttributeHelper.createThemedTextStyle(context);
    textStyle.setShouldLayoutEmptyText(true);
    textStyle.setHighlightColor(Color.TRANSPARENT);
    return textStyle;
  }

  public static void maybeSetTextAlignment(final TextStyle textStyle, final Integer textAlign) {
    if (textAlign != null) {
      final TextAlignment textAlignment;
      switch (textAlign) {
        case Gravity.CENTER_HORIZONTAL:
          textAlignment = TextAlignment.CENTER;
          break;
        case Gravity.START:
          textAlignment = TextAlignment.TEXT_START;
          break;
        case Gravity.END:
          textAlignment = TextAlignment.TEXT_END;
          break;
        default:
          textAlignment = TextAlignment.TEXT_START;
      }
      textStyle.setAlignment(textAlignment);
    }
  }

  public static void maybeSetTextDirection(
      final TextStyle textStyle, final @Nullable String textDirectionStr, final boolean isRTL) {
    if (textDirectionStr != null) {
      final TextDirectionHeuristicCompat textDirection;

      switch (textDirectionStr) {
        case "device_locale":
          textDirection = TextDirectionHeuristicsCompat.LOCALE;
          break;
        case "text_first_strong":
          textDirection =
              isRTL
                  ? TextDirectionHeuristicsCompat.FIRSTSTRONG_RTL
                  : TextDirectionHeuristicsCompat.FIRSTSTRONG_LTR;
          break;
        default:
          textDirection = TextDirectionHeuristicsCompat.LOCALE;
      }
      textStyle.setTextDirection(textDirection);
    }
  }

  public static void maybeSetLineHeight(final TextStyle textStyle, final float lineHeight) {
    if (lineHeight >= 0) {
      textStyle.setLineHeight(lineHeight);
    }
  }

  public static void maybeSetLineHeightMultiplier(
      final TextStyle textStyle, final float lineHeightMultiplier) {
    maybeSetLineHeightMultiplier(textStyle, lineHeightMultiplier, true);
  }

  public static void maybeSetMaxNumberOfLines(
      final TextStyle textStyle, final int maxNumberOfLines) {
    if (maxNumberOfLines > -1) {
      textStyle.setMaxLines(maxNumberOfLines);
      textStyle.setEllipsize(TextUtils.TruncateAt.END);
    }
  }

  public static void maybeSetLineHeightMultiplier(
      final TextStyle textStyle, final float lineHeightMultiplier, boolean applyToFirstLine) {
    if (lineHeightMultiplier > 0) {
      // This is a special behavior to match Bloks classic and the iOS behavior.
      textStyle.setShouldAddSpacingExtraToFirstLine(applyToFirstLine);
      textStyle.setLineHeightMultiplier(lineHeightMultiplier);
    }
  }

  public void setTextSize(int pixels) {
    this.textSize = pixels;
  }

  public void setTypeface(Typeface typeface) {
    this.typeface = typeface;
  }

  public void setShouldAddSpacingExtraToFirstLine(boolean shouldAddSpacingExtraToFirstLine) {
    this.shouldAddSpacingExtraToFirstLine = shouldAddSpacingExtraToFirstLine;
  }

  public void setLineSpacingExtra(float lineSpacingExtra) {
    this.lineSpacingExtra = lineSpacingExtra;
  }

  public void setLineHeight(float lineHeight) {
    this.lineHeight = lineHeight;
  }

  public void setLineHeightMultiplier(Float lineHeightMultiplier) {
    this.lineHeightMultiplier = lineHeightMultiplier;
  }

  public void setAlignment(TextAlignment textAlignment) {
    this.alignment = textAlignment;
  }

  public void setTextDirection(TextDirectionHeuristicCompat textDirection) {
    this.textDirection = textDirection;
  }

  public void setVerticalGravity(VerticalGravity verticalGravity) {
    this.verticalGravity = verticalGravity;
  }

  public void setMaxLines(int maxNumberOfLines) {
    this.maxLines = maxNumberOfLines;
  }

  public void setHighlightColor(int color) {
    this.highlightColor = color;
  }

  public void setHighlightCornerRadius(int radius) {
    this.highlightCornerRadius = radius;
  }

  public void setEllipsize(TextUtils.TruncateAt end) {
    this.ellipsize = end;
  }

  public void setTextStyle(int textStyle) {
    this.textStyle = textStyle;
  }

  public void setSingleLine(boolean singleLine) {
    this.isSingleLine = singleLine;
  }

  public void setTextColor(@ColorInt int textColor) {
    this.textColor = textColor;
    this.textColorStateList = null;
  }

  public void setTextColorStateList(ColorStateList textColorStateList) {
    this.textColorStateList = textColorStateList;
    this.textColor = 0;
  }

  public void setCustomEllipsisText(CharSequence customEllipsisText) {
    this.customEllipsisText = customEllipsisText;
  }

  public void setIncludeFontPadding(boolean includeFontPadding) {
    this.includeFontPadding = includeFontPadding;
  }

  public void setExtraSpacingLeft(float extraSpacingLeft) {
    this.extraSpacingLeft = extraSpacingLeft;
  }

  public void setExtraSpacingRight(float extraSpacingRight) {
    this.extraSpacingRight = extraSpacingRight;
  }

  public void setShadowRadius(float shadowRadius) {
    this.shadowRadius = shadowRadius;
  }

  public void setShadowDx(float shadowDx) {
    this.shadowDx = shadowDx;
  }

  public void setShadowDy(float shadowDy) {
    this.shadowDy = shadowDy;
  }

  public void setShadowColor(int shadowColor) {
    this.shadowColor = shadowColor;
  }

  public void setRoundedBackgroundColor(RoundedBackgroundProps roundedBackgroundProps) {
    this.roundedBackgroundProps = roundedBackgroundProps;
  }

  public static class RoundedBackgroundProps {
    final RectF padding;
    final float cornerRadius;
    final @ColorInt int backgroundColor;

    public RoundedBackgroundProps(RectF padding, float cornerRadius, int backgroundColor) {
      this.padding = padding;
      this.cornerRadius = cornerRadius;
      this.backgroundColor = backgroundColor;
    }
  }
}

File: litho-rendercore-text/src/main/java/com/facebook/rendercore/text/TextStyle.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.rendercore.text

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.RectF
import android.graphics.Typeface
import android.text.TextUtils
import android.view.Gravity
import androidx.annotation.ColorInt
import androidx.core.text.TextDirectionHeuristicCompat
import androidx.core.text.TextDirectionHeuristicsCompat
import kotlin.jvm.JvmField

class TextStyle : Cloneable {

  @JvmField var ellipsize: TextUtils.TruncateAt? = null
  @JvmField var customEllipsisText: CharSequence? = null
  @JvmField var includeFontPadding: Boolean = true
  @JvmField var minLines: Int = Int.MIN_VALUE
  @JvmField var maxLines: Int = Int.MAX_VALUE
  @JvmField var minEms: Int = UNSET
  @JvmField var maxEms: Int = UNSET
  @JvmField var minTextWidth: Int = 0
  @JvmField var maxTextWidth: Int = Int.MAX_VALUE
  @JvmField var shadowRadius: Float = 0f
  @JvmField var shadowDx: Float = 0f
  @JvmField var shadowDy: Float = 0f
  @ColorInt @JvmField var shadowColor: Int = Color.GRAY
  @JvmField var isSingleLine: Boolean = false
  @ColorInt @JvmField var linkColor: Int = Color.BLUE
  @JvmField var textSize: Int = UNSET
  @JvmField var shouldAddSpacingExtraToFirstLine: Boolean = false
  @JvmField var lineSpacingExtra: Float = 0f
  @JvmField var lineHeight: Float = Float.MAX_VALUE
  @JvmField var lineHeightMultiplier: Float = 1f
  @JvmField var letterSpacing: Float = 0f
  @JvmField var textStyle: Int = Typeface.DEFAULT.style
  @JvmField var typeface: Typeface? = null
  @JvmField var alignment: TextAlignment = TextAlignment.TEXT_START
  @JvmField var breakStrategy: Int = UNSET
  @JvmField var hyphenationFrequency: Int = 0
  @JvmField var justificationMode: Int = 0
  @JvmField var textDirection: TextDirectionHeuristicCompat? = null
  @JvmField var clipToBounds: Boolean = true
  @JvmField var verticalGravity: VerticalGravity = VerticalGravity.TOP
  @ColorInt @JvmField var highlightColor: Int = Color.TRANSPARENT
  @ColorInt @JvmField var keyboardHighlightColor: Int = Color.TRANSPARENT
  @JvmField var highlightStartOffset: Int = UNSET
  @JvmField var highlightEndOffset: Int = UNSET
  @JvmField var highlightCornerRadius: Int = 0
  @JvmField var shouldLayoutEmptyText: Boolean = false
  @JvmField var minimallyWide: Boolean = false
  @JvmField var minimallyWideThreshold: Int = 0

  /**
   * Click offset amount to determine how far off the ClickableSpan bounds user can click to be able
   * to trigger ClickableSpan's click action. This could be useful in a densely lined text with
   * links like 'Continue reading ...' in NewsFeed to be able to click that easily.
   */
  @JvmField var clickableSpanExpandedOffset: Float = 0f
  @JvmField var shouldTruncateTextUsingConstraints: Boolean = false
  @JvmField var manualBaselineSpacing: Int = Int.MIN_VALUE
  @JvmField var manualCapSpacing: Int = Int.MIN_VALUE
  @JvmField var extraSpacingLeft: Float = 0f
  @JvmField var extraSpacingRight: Float = 0f
  @JvmField var outlineWidth: Float = 0f
  @ColorInt @JvmField var outlineColor: Int = Color.TRANSPARENT
  @JvmField var roundedBackgroundProps: RoundedBackgroundProps? = null
  @JvmField var accessibilityLabel: String? = null
  @JvmField var truncationStyle: TruncationStyle = TruncationStyle.USE_MAX_LINES

  @ColorInt private var _textColor: Int = Color.BLACK
  private var _textColorStateList: ColorStateList? = null

  var textColor: Int
    @ColorInt get() = _textColor
    set(@ColorInt value) {
      _textColor = value
      _textColorStateList = null
    }

  var textColorStateList: ColorStateList?
    get() = _textColorStateList
    set(value) {
      _textColorStateList = value
      _textColor = 0
    }

  fun makeCopy(): TextStyle {
    try {
      return super.clone() as TextStyle
    } catch (e: CloneNotSupportedException) {
      throw RuntimeException(e)
    }
  }

  class RoundedBackgroundProps(
      @JvmField val padding: RectF,
      @JvmField val cornerRadius: Float,
      @field:ColorInt @JvmField val backgroundColor: Int
  )

  companion object {
    const val UNSET: Int = -1

    @JvmStatic
    fun createDefaultConfiguredTextStyle(context: Context): TextStyle =
        TextStylesAttributeHelper.createThemedTextStyle(context).apply {
          shouldLayoutEmptyText = true
          highlightColor = Color.TRANSPARENT
        }

    @JvmStatic
    fun maybeSetTextAlignment(textStyle: TextStyle, textAlign: Int?) {
      if (textAlign != null) {
        val textAlignment =
            when (textAlign) {
              Gravity.CENTER_HORIZONTAL -> TextAlignment.CENTER
              Gravity.START -> TextAlignment.TEXT_START
              Gravity.END -> TextAlignment.TEXT_END
              else -> TextAlignment.TEXT_START
            }
        textStyle.alignment = textAlignment
      }
    }

    @JvmStatic
    fun maybeSetTextDirection(textStyle: TextStyle, textDirectionStr: String?, isRTL: Boolean) {
      if (textDirectionStr != null) {
        val textDirection =
            when (textDirectionStr) {
              "device_locale" -> TextDirectionHeuristicsCompat.LOCALE
              "text_first_strong" ->
                  if (isRTL) TextDirectionHeuristicsCompat.FIRSTSTRONG_RTL
                  else TextDirectionHeuristicsCompat.FIRSTSTRONG_LTR

              else -> TextDirectionHeuristicsCompat.LOCALE
            }
        textStyle.textDirection = textDirection
      }
    }

    @JvmStatic
    fun maybeSetLineHeight(textStyle: TextStyle, lineHeight: Float) {
      if (lineHeight >= 0) {
        textStyle.lineHeight = lineHeight
      }
    }

    @JvmStatic
    fun maybeSetMaxNumberOfLines(textStyle: TextStyle, maxNumberOfLines: Int) {
      if (maxNumberOfLines > -1) {
        textStyle.maxLines = maxNumberOfLines
        textStyle.ellipsize = TextUtils.TruncateAt.END
      }
    }

    @JvmStatic
    @JvmOverloads
    fun maybeSetLineHeightMultiplier(
        textStyle: TextStyle,
        lineHeightMultiplier: Float,
        applyToFirstLine: Boolean = true
    ) {
      if (lineHeightMultiplier > 0) {
        // This is a special behavior to match Bloks classic and the iOS behavior.
        textStyle.shouldAddSpacingExtraToFirstLine = applyToFirstLine
        textStyle.lineHeightMultiplier = lineHeightMultiplier
      }
    }

    @JvmStatic
    fun addThemedTextStyleForContextToCache(context: Context) {
      TextStylesAttributeHelper.addThemedTextStyleForContext(context)
    }
  }
}
